File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/widget/MensaRemoteViewFactory.java
package de.tum.in.tumcampusapp.component.ui.cafeteria.widget;

import android.content.Context;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.ui.cafeteria.controller.CafeteriaManager;
import de.tum.in.tumcampusapp.component.ui.cafeteria.model.CafeteriaMenu;
import de.tum.in.tumcampusapp.component.ui.cafeteria.model.CafeteriaPrices;

public class MensaRemoteViewFactory implements RemoteViewsService.RemoteViewsFactory {

    private static final Pattern COMPILE = Pattern.compile("\\([^\\)]+\\)");

    private final Context mApplicationContext;
    private List<CafeteriaMenu> mMenus = new ArrayList<>();

    public MensaRemoteViewFactory(Context context) {
        this.mApplicationContext = context;
    }

    @Override
    public void onCreate() {
        CafeteriaManager mensaManager = new CafeteriaManager(mApplicationContext);
        mMenus = mensaManager.getBestMatchCafeteriaMenus();
    }

    @Override
    public void onDataSetChanged() {
        // Noop
    }

    @Override
    public void onDestroy() {
        // Noop 
    }

    @Override
    public int getCount() {
        return mMenus.size();
    }

    @Override
    public RemoteViews getViewAt(int position) {
        CafeteriaMenu currentItem = mMenus.get(position);

        RemoteViews rv = new RemoteViews(mApplicationContext.getPackageName(), R.layout.mensa_widget_item);

        String menuContent = COMPILE.matcher(currentItem.getName())
                                    .replaceAll("")
                                    .trim();
        String menuText = mApplicationContext.getString(
                R.string.menu_with_long_type_format_string, menuContent, currentItem.getTypeLong());
        rv.setTextViewText(R.id.menu_content, menuText);

        String price = CafeteriaPrices.INSTANCE.getPrice(mApplicationContext, currentItem.getTypeLong());
        if (price != null) {
            rv.setTextViewText(R.id.menu_price, price + " €");
        }

        return rv;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/widget/MensaRemoteViewFactory.kt
package de.tum.`in`.tumcampusapp.component.ui.cafeteria.widget

import android.content.Context
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.controller.CafeteriaManager
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.model.CafeteriaMenu
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.model.deserialization.DishPrices
import java.util.*

class MensaRemoteViewFactory(private val applicationContext: Context) : RemoteViewsService.RemoteViewsFactory {
    private var menus: List<CafeteriaMenu> = ArrayList()

    override fun onCreate() {
        menus = CafeteriaManager(applicationContext).bestMatchCafeteriaMenus
    }

    override fun onDataSetChanged() { /* Noop */ }

    override fun onDestroy() { /* Noop */ }

    override fun getCount() = menus.size

    override fun getViewAt(position: Int): RemoteViews {
        if (position >= menus.size) {
            // No idea why this happens, but getViewAt is occasionally called with position == size
            return loadingView
        }

        val (_, _, _, _, dishType, name, _, _) = menus[position]
        val remoteViews = RemoteViews(applicationContext.packageName, R.layout.mensa_widget_item)

        val menuText = applicationContext.getString(
                R.string.menu_with_long_type_format_string, name, dishType)
        remoteViews.setTextViewText(R.id.menu_content, menuText)

        DishPrices.getPrice(applicationContext, dishType)?.let {
            remoteViews.setTextViewText(R.id.menu_price, "$it €")
        }

        return remoteViews
    }

    override fun getLoadingView() = RemoteViews(applicationContext.packageName, R.layout.mensa_widget_loading_item)

    override fun getViewTypeCount() = 1

    override fun getItemId(position: Int) = position.toLong()

    override fun hasStableIds() = true
}
